wire-schema/src/main/java/com/squareup/wire/schema/SchemaLoader.java

```diff
@@ -0,0 +1,155 @@
+  public SchemaLoader addDirectory(Path path) {
+    directories.add(path);
+    return this;
+  }
+
+  /** Add directories from which proto files will be loaded. */
+  public SchemaLoader addDirectories(Iterable<? extends Path> paths) {
+    directories.addAll(paths);
+    return this;
+  }
+
+  /**
+   * Add one or more proto files to load. Dependencies will be loaded automatically from the
+   * configured directories.
+   */
+  public SchemaLoader addProto(File file) {
+    return addProto(file.toPath());
+  }
+
+  /**
+   * Add proto files to load. Dependencies will be loaded automatically from the configured
+   * directories.
+   */
+  public SchemaLoader addProtos(Collection<? extends File> files) {
+    return addProtos(Iterables.transform(files, FILE_TO_PATH));
+  }
+
+  /**
+   * Add one or more proto files to load. Dependencies will be loaded automatically from the
+   * configured directories.
+   */
+  public SchemaLoader addProto(Path path) {
+    protos.add(path);
+    return this;
+  }
+
+  /**
+   * Add proto files to load. Dependencies will be loaded automatically from the configured
+   * directories.
+   */
+  public SchemaLoader addProtos(Iterable<? extends Path> paths) {
+    protos.addAll(paths);
+    return this;
+  }
+
+  public Schema load() throws IOException {
+    ImmutableList<Path> directories = this.directories.build();
+    if (directories.isEmpty()) {
+      throw new IllegalStateException("No directories added.");
+    }
+
+    Deque<Path> protos = new ArrayDeque<>(this.protos.build());
+    if (protos.isEmpty()) {
+      // TODO traverse all files in every directory.
+    }
+
+    Map<Path, ProtoFile> loaded = new LinkedHashMap<>();
+    while (!protos.isEmpty()) {
+      Path proto = protos.removeFirst();
+      if (loaded.containsKey(proto)) {
+        continue;
+      }
+
+      ProtoFileElement element = null;
+      for (Path directory : directories) {
+        if (proto.isAbsolute() && !proto.startsWith(directory)) {
+          continue;
+        }
+        Path resolvedPath = directory.resolve(proto);
+        if (Files.exists(resolvedPath)) {
+          Location location = Location.get(directory.toString(), proto.toString());
+          try (Source source = Okio.source(resolvedPath)) {
+            String data = Okio.buffer(source).readUtf8();
+            element = ProtoParser.parse(location, data);
+          } catch (IOException e) {
+            throw new IOException("Failed to load " + proto, e);
+          }
+          break;
+        }
+      }
+      if (element == null) {
+        throw new FileNotFoundException("Failed to locate " + proto + " in " + directories);
+      }
+
+      ProtoFile protoFile = ProtoFile.get(element);
+      loaded.put(proto, protoFile);
+
+      // Queue dependencies to be loaded.
+      FileSystem fs = proto.getFileSystem();
+      for (String importPath : element.imports()) {
+        protos.addLast(fs.getPath(importPath));
```

Comment:
Unsure the consequences of specifying a file system here. It shouldn't be file-system specific.

Actually not sure about using `Path` at all for the path part, since we also want to support `.jar` files in what is currently the directories list.